### PR TITLE
feat(shared): `tbl_deep_extend` improvements

### PR DESCRIPTION
**Pull request by folke**
Right now, empty tables are treated as lists in the deep extend method, which means if you use `force` and have an empty table value in a table on the right, it will overwrite the existsing value.

`vim.tbl_deep_extend("force", { ss = { a = 1 } }, { ss = {} })` ==> empty
`vim.tbl_deep_extend("force", { ss = { a = 1 } }, { ss = { b = 1} })` => has both a and b

It also fixes an issue when merging values of a different type. This used to fail:
`vim.tbl_deep_extend("force", {a=true}, {a={b= 1}})`

This PR fixes that.
Additional tests have been included

### CODE REVIEW

**Comment by mjlbach** [comment]
@folke it would probably be useful to state here the utility of vim.merge for future reference. Also note, we'll need to update https://github.com/neovim/neovim/issues/14090 with the changes to tbl_deep_extend.

@tjdevries are you ok with vim.merge? It's a top level namespace function so I'm not comfortable merging it without a sign-off. The implementation looks correct to me.

**Comment by folke** [comment]
@mjlbach the existing tests for `tbl_deep_extend` all passed, but I'll add in a couple of new ones that should now pass with the fixes and would've failed previously.

I'll also add some tests for `vim.merge`.

**Comment by mjlbach** [comment]
Yes, that would be great. I tested all of them myself when reviewing too. We might want to expand the docstring for deep_extend too.

**Comment by folke** [comment]
@mjlbach added tests and updated the description of the PR.

File: runtime/lua/vim/shared.lua
Comment (mfussenegger) [inline runtime/lua/vim/shared.lua:204]:
```suggestion
  return type(v) == "table" and #v == 0
```

`type(v) == 'table'` would already imply that `v ~= nil`

File: runtime/lua/vim/shared.lua
Comment (mfussenegger) [inline runtime/lua/vim/shared.lua:245]: Is it worth mentioning that this may not be safe to use for dict-style tables where numbers are used as key?

E.g a common pattern is to have the buf number as key - then you could have something like this:

```
vim.merge({ [10] = 'a'}, { [20] = 'b'})
```

Which appears to work, but if you have a `[1] = ...` entry as well, it will override the left value.

```
vim.merge({ [1] = 'a', [10] = 'a'}, { [20] = 'b'})
```

File: runtime/lua/vim/shared.lua
Comment (folke) [reply runtime/lua/vim/shared.lua:204]: that was me trying to debug a segmentation fault. turned out it was tressitter (parsers were built with different libs). will change again

File: runtime/lua/vim/shared.lua
Comment (folke) [reply runtime/lua/vim/shared.lua:245]: Good catch!

This also impacts `tbl_deep_extend`. In there lists always overwrite the left-hand-side, but with my change that would no longer be the case. Will make sure it follows the old behavior.

Right now, `vim.tbl_islist({[10] = 'a'})` returns `true`, but that doesn't seem right? (that was the check that used to be there before)

Not sure what the wanted behavior should be here.

I'm also a bit hesitant to change the behavior of `tbl_islist`.

For `vim.merge(`, I can change it of course.

Maybe we should do the `islist` check there, by checking that all keys are consecutive from 1 to total count? (min key = 1 and max key = count of keys and all keys are numbers)?

This way, your second example would also merge properly. but:
`vim.merge({ [1] = 'a'}, { [20] = 'b'})` would still remove the `1`.

Another approach could be that we merge if one of the params is `can_merge` and the other must be at least a `table`?

That way, merging two list-like tables, will overwrite each other, while your example and my other example above here, would still merge.

Wdyt?

Comment (folke) [reply runtime/lua/vim/shared.lua:245]: I don't think it's a good idea to try being smart about this. I changed the implementation back to how it was for tbl_deep_extend. The same check is now used for `vim.merge`.

We will only merge tables if either the table is empty, or it is not `vim.tbl_islist`.

For your examples, this means:

```lua
vim.merge({ [10] = 'a'}, { [20] = 'b'}) -- result: {[20] = 'b'}

vim.merge({ [1] = 'a', [10] = 'a'}, { [20] = 'b'}) -- result: {[20] = 'b'}
```

**Comment by folke** [comment]
I just pushed an update that restores the behavior with tables with number indices as it used to be.

So tables can be merged, if both are either an empty table, or not `vim.tbl_islist`

**Comment by mjlbach** [comment]
Lets break vim.merge into its own PR and merge the deep_extend changes.

**Comment by folke** [comment]
@mjlbach I removed `vim.merge` from this PR

**Comment by folke** [comment]
`vim.merge` is now at #15124

**Review by mjlbach — APPROVED** [review]

**Comment by mjlbach** [comment]
@folke can you update https://github.com/neovim/neovim/issues/14090? 

Thanks!

**Comment by mjlbach** [comment]
Actually before I merge can you updated the commit body with a few examples of the old and new behavior?

**Comment by folke** [comment]
@mjlbach I updates the commit messages.

I can't update #14090 since it's locked.

**Comment by mjlbach** [comment]
We should let all org members write to locked issues, not sure how to change that setting but I think it has to be someone with admin rights to the repo.

**Pull request merged**
